tsl/src/compression/compression.c

```diff
@@ -530,6 +555,205 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
 static void segment_info_update(SegmentInfo *segment_info, Datum val, bool is_null);
 static bool segment_info_datum_is_in_group(SegmentInfo *segment_info, Datum datum, bool is_null);
 
+/* Find segment by index for setting the correct sequence number if
+ * we are trying to roll up chunks while compressing
+ */
+static Oid
+get_compressed_chunk_index(Relation compressed_chunk, int16 *uncompressed_col_to_compressed_col,
+						   PerColumn *per_column, int n_input_columns)
+{
+	ListCell *lc;
+	int i;
+
+	List *index_oids = RelationGetIndexList(compressed_chunk);
+
+	foreach (lc, index_oids)
+	{
+		Oid index_oid = lfirst_oid(lc);
+		bool matches = true;
+		int num_segmentby_columns = 0;
+		Relation index_rel = index_open(index_oid, AccessShareLock);
+		IndexInfo *index_info = BuildIndexInfo(index_rel);
+
+		for (i = 0; i < n_input_columns; i++)
+		{
+			if (per_column[i].segmentby_column_index < 1)
+				continue;
+
+			/* Last member of the index must be the sequence number column. */
+			if (per_column[i].segmentby_column_index >= index_rel->rd_att->natts)
+			{
+				matches = false;
+				break;
+			}
+
+			int index_att_offset = AttrNumberGetAttrOffset(per_column[i].segmentby_column_index);
+
+			if (index_info->ii_IndexAttrNumbers[index_att_offset] !=
+				AttrOffsetGetAttrNumber(uncompressed_col_to_compressed_col[i]))
+			{
+				matches = false;
+				break;
+			}
+
+			num_segmentby_columns++;
+		}
+
+		/* Check that we have the correct number of index attributes
+		 * and that the last one is the sequence number
+		 */
+		if (num_segmentby_columns != index_rel->rd_att->natts - 1 ||
+			0 != namestrcmp((Name) &index_rel->rd_att->attrs[num_segmentby_columns].attname,
+							COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME))
+			matches = false;
+
+		index_close(index_rel, AccessShareLock);
+
+		if (matches)
+			return index_oid;
+	}
+
+	return InvalidOid;
+}
+
+/* Scan index to get the sequence number for current group.
+ * This is necessary to do when merging chunks. If the chunk is empty,
+ * index will always return 0 and the sequence number will start from
+ * SEQUENCE_NUM_GAP.
+ */
+static int32
+get_sequence_number_for_current_group(Relation table_rel, Oid index_oid,
+									  int16 *uncompressed_col_to_compressed_col,
+									  PerColumn *per_column, int n_input_columns)
+{
+	/* If no index, relation must be empty so just reset. */
+	if (index_oid == InvalidOid)
+		return SEQUENCE_NUM_GAP;
```

Comment:
That is not a correct assumption, the presence of an index is not dependent on there being data and vice versa.